pkg/pdp/scheduler/engine.go

```diff
@@ -0,0 +1,362 @@
+		h := &taskTypeHandler{
+			TaskInterface:   impl,
+			TaskTypeDetails: impl.TypeDetails(),
+			TaskEngine:      e,
+		}
+		e.handlers = append(e.handlers, h)
+		e.taskMap[h.TaskTypeDetails.Name] = h
+
+		// Start the adder routine for the task type.
+		go h.Adder(h.AddTask)
+
+		// **Start the periodic "bored" routine if provided**
+		if h.TaskTypeDetails.IAmBored != nil {
+			go func(h *taskTypeHandler) {
+				err := h.TaskTypeDetails.IAmBored(h.AddTask)
+				if err != nil {
+					log.Warnf("IAmBored for task %s returned error: %v",
+						h.TaskTypeDetails.Name, err)
+				}
+			}(h)
+		}
+	}
+
+	go e.poller()
+
+	return e, nil
+}
+
+// GracefullyTerminate stops new task scheduling.
+func (e *TaskEngine) GracefullyTerminate() {
+	e.cancel()
+	// Optionally wait for currently running tasks to finish.
+}
+
+// poller continuously checks for work.
+func (e *TaskEngine) poller() {
+	pollDuration := 3 * time.Second
+	pollNextDuration := 100 * time.Millisecond
+	nextWait := pollNextDuration
+
+	for {
+		select {
+		case <-time.After(nextWait):
+		case <-e.ctx.Done():
+			return
+		}
+		nextWait = pollDuration
+
+		accepted := e.pollerTryAllWork()
+		if accepted {
+			nextWait = pollNextDuration
+		}
+		// Here you could also call a follow-up work routine if needed.
+	}
+}
+
+const (
+	WorkSourcePoller   = "poller"
+	WorkSourceRecover  = "recovered"
+	WorkSourceIAmBored = "bored"
+)
+
+// pollerTryAllWork looks for unassigned tasks in the DB and schedules them.
+func (e *TaskEngine) pollerTryAllWork() bool {
+	// Optional cleanup logic.
+	if time.Since(e.lastCleanup) > 5*time.Minute {
+		e.lastCleanup = time.Now()
+		// (Cleanup code can be added here if needed.)
+	}
+
+	// Iterate over all registered task types.
+	for _, h := range e.handlers {
+		var tasks []models.Task
+		// Fetch tasks for this type that are unassigned.
+		// (Assuming unassigned tasks are filtered by name; add additional conditions if needed.)
+		if err := e.db.WithContext(e.ctx).
+			Where("name = ? AND owner_id IS NULL", h.TaskTypeDetails.Name).
+			Order("update_time").
+			Find(&tasks).Error; err != nil {
+			log.Errorf("Unable to read work for task type %s: %v", h.TaskTypeDetails.Name, err)
+			continue
+		}
+
+		var taskIDs []TaskID
+		// Filter tasks based on retry logic.
+		// Since the Task model no longer has a retries field, we assume a default value (e.g., 0)
+		// or adjust the logic if you retrieve retries from another source.
+		for _, t := range tasks {
+			if h.TaskTypeDetails.RetryWait == nil ||
+				time.Since(t.UpdateTime) > h.TaskTypeDetails.RetryWait(0) {
+				taskIDs = append(taskIDs, TaskID(t.ID))
+			}
+		}
+
+		if len(taskIDs) > 0 {
+			accepted := h.considerWork(WorkSourcePoller, taskIDs, e.db)
+			if accepted {
+				return true
+			}
+			log.Warnf("Work not accepted for %d %s task(s)", len(taskIDs), h.TaskTypeDetails.Name)
+		}
+	}
+
+	// if no work was accepted, are we bored? Then find work in priority order.
+	for _, v := range e.handlers {
```

Comment:
checking logic: this only runs if all the other tasks don't go through? but also we have some tasks with manual adds and somewhere tasks only get added via the I am bored function